pennylane/qchem/hartree_fock.py

```diff
@@ -126,16 +126,36 @@ def _scf(*args):
         r = mol.coordinates
         n_electron = mol.n_electrons
 
-        if r.requires_grad:
-            args_r = [[args[0][i]] * mol.n_basis[i] for i in range(len(mol.n_basis))]
-            args_ = [*args] + [qml.math.vstack(list(itertools.chain(*args_r)))]
-            rep_tensor = repulsion_tensor(basis_functions)(*args_[1:])
-            s = overlap_matrix(basis_functions)(*args_[1:])
-            h_core = core_matrix(basis_functions, charges, r)(*args_)
+        if qml.math.get_interface(r) == "autograd":
+            if getattr(r, "requires_grad", False):
+                args_r = [[args[0][i]] * mol.n_basis[i] for i in range(len(mol.n_basis))]
+                # In autograd, _scf re-orders args from r, alpha, coeff to args_ = r, alpha, coeff, r
+                args_ = [*args] + [qml.math.vstack(list(itertools.chain(*args_r)))]
+                rep_tensor = repulsion_tensor(basis_functions)(*args_[1:])
+                s = overlap_matrix(basis_functions)(*args_[1:])
+                h_core = core_matrix(basis_functions, charges, r)(*args_)
+            else:
+                rep_tensor = repulsion_tensor(basis_functions)(*args)
+                s = overlap_matrix(basis_functions)(*args)
+                h_core = core_matrix(basis_functions, charges, r)(*args)
         else:
-            rep_tensor = repulsion_tensor(basis_functions)(*args)
-            s = overlap_matrix(basis_functions)(*args)
-            h_core = core_matrix(basis_functions, charges, r)(*args)
+            # NOTE: In JAX, we want to keep the ordering as r, coeff, alpha.
+            # But for core_matrix, we pass in r, r, coeff, alpha.
+            if (
+                len(args) > 0
+                and qml.math.get_interface(r) == "jax"
+                and qml.math.requires_grad(args[0])
+            ):
+                args_r = [[args[0][i]] * mol.n_basis[i] for i in range(len(mol.n_basis))]
+                args_ = [*args] + [qml.math.vstack(list(itertools.chain(*args_r)))]
+                args_ = [args_[0], args_[3], args_[1], args_[2]]
```

Comment:
I guess technically, only this line makes use of the "jax" if-statement, so I guess we could contain only this in it.